### PR TITLE
xl: Avoid called getdents for folders in listing

### PR DESCRIPTION
**Pull request by vadmeste**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request, I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/xl-storage_test.go [hunk 1]
```diff
@@ -216,6 +217,7 @@ func TestXLStorageIsDirEmpty(t *testing.T) {
 		t.Error("expected true for empty dir, got false")
 	}
 }
+*/
```
Comment (harshavardhana) [inline cmd/xl-storage_test.go]:
```suggestion
```

File: cmd/xl-storage_test.go [hunk 2]
```diff
@@ -185,6 +185,7 @@ func TestXLStorageGetDiskInfo(t *testing.T) {
 	}
 }
 
+/*
```
Comment (harshavardhana) [inline cmd/xl-storage_test.go]:
```suggestion
```

File: cmd/xl-storage.go [hunk 3]
```diff
@@ -194,8 +194,15 @@ func getValidPath(path string) (string, error) {
 	return path, nil
 }
 
-// isDirEmpty - returns whether given directory is empty or not.
-func isDirEmpty(dirname string) bool {
+// dirWithoutSubdirs - returns whether given directory have other directories or not.
+func dirWithoutSubdirs(dirname string) bool {
```
Comment (harshavardhana) [inline cmd/xl-storage.go]:
```suggestion
func isDirEmpty(dirname string) bool {
```

File: cmd/metacache-walk.go [hunk 4]
```diff
@@ -354,7 +354,7 @@ func (s *xlStorage) WalkDir(ctx context.Context, opts WalkDirOptions, wr io.Writ
 				// NOT an object, append to stack (with slash)
 				// If dirObject, but no metadata (which is unexpected) we skip it.
 				if !isDirObj {
-					if !isDirEmpty(pathJoinBuf(sb, volumeDir, meta.name)) {
+					if !dirWithoutSubdirs(pathJoinBuf(sb, volumeDir, meta.name)) {
```
Comment (harshavardhana) [inline cmd/metacache-walk.go]:
```suggestion
					if !isDirEmpty(pathJoinBuf(sb, volumeDir, meta.name)) {
```

File: cmd/xl-storage.go [hunk 5]
```diff
@@ -194,8 +194,15 @@ func getValidPath(path string) (string, error) {
 	return path, nil
 }
 
-// isDirEmpty - returns whether given directory is empty or not.
-func isDirEmpty(dirname string) bool {
+// dirWithoutSubdirs - returns whether given directory have other directories or not.
```
Comment (harshavardhana) [inline cmd/xl-storage.go]:
```suggestion
// isDirEmpty - returns whether the given directory has subdirectories or not. 
// this function is mainly meaningful for MinIO backend format, this is not
// a generic isDirEmpty() function.
```

**Review by klauspost — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by kevinastone** [comment]
This change breaks when the filesystem is a `cifs` mount.  Stat.NLinks is always 2 despite the dir not being empty.

**Comment by harshavardhana** [comment]
> This change breaks when the filesystem is a `cifs` mount. Stat.NLinks is always 2 despite the dir not being empty

Please open new issues if you have something to report. Avoid making comments on merged PRs.

**Pull request merged**